### PR TITLE
twister: add sysbuild support

### PR DESCRIPTION
**Pull request by danieldegrasse**
This PR adds support for sysbuild to twister. This will enable test cases that require multiple images, such as verifying MCUBoot support, or running tests on the second core of some dual core systems. The changes are designed to be as unintrusive as possible, so sysbuild can be used with existing projects with no impact.

Twister can be instructed to use sysbuild via the `sysbuild` YAML tag, and will then build the project with sysbuild. By default, no external sysbuild projects will be compiled in, but they can be enabled via the `extra_args` YAML tag for tests.

Enabling sysbuild will not affect twister's filtering options, and when multiple images are included into the build sysbuild will parse Kconfig and devicetree values from the default sysbuild image (the main application being build, that is not added via `ExternalZephyrProject_Add()`). This PR does not enable twister to filter based on configuration settings in the external sysbuild images.

This PR also adds an MCUBoot test, which will verify that twister sysbuild support is working correctly, as well as testing MCUBoot support. This test is currently enabled for the `mimxrt1060_evk` and `frdm_k64f`

### CODE REVIEW

**Comment by danieldegrasse** [comment]
@tejlmand FYI

File: scripts/pylib/twister/twisterlib/runner.py
Comment (tejlmand) [inline scripts/pylib/twister/twisterlib/runner.py:331]: do we really need to do that ?
The `self.source_dir` should be the default sysbuild application so any special reason that cannot be used here or that it is undesired to use it ?

I can envision future complex scenarios where we want to use sysbuild for multiple samples where default makes less sense, but I don't beleive that's the case here.

File: scripts/twister
Comment (tejlmand) [inline scripts/twister:49]: if this is just temporary in order to limit the scope, then fine, but long term we should be able to still allow Kconfig or dt filtering and just use the `.config` or dt from the main sample build folder instead.

But I can accept the limitation for now.

**Review by tejlmand — COMMENTED** [review]
great to see this 🎉 

Couple of questions before +1.

File: scripts/twister
Comment (danieldegrasse) [reply scripts/twister:49]: I need to update this, thank you for pointing it out. KConfig and devicetree filtering are supported, just only on the main systembuild application

File: scripts/pylib/twister/twisterlib/runner.py
Comment (danieldegrasse) [reply scripts/pylib/twister/twisterlib/runner.py:331]: We do I believe, because twister does not parse the source application, it parses the generated `.config` and `edt.pickle` files in the build folder for an application. This leaves us with the following options:

- Don't parse `domains.yaml`, and simply search the root of the build folder: This will result in searching the "meta" sysbuild application, that includes all other applications in the build as external projects. This will work, but we won't be able to do useful Kconfig or devicetree filtering
- Parse the `domains.yaml` and select the default application's build directory: This will enable Kconfig and devicetree filtering on the primary application only. It's the solution I have enabled currently due to the low complexity and impact
- Parse the `domains.yaml` and generate filter data for all applications: This is probably the best solution long term, but I avoided it in this PR because I wasn't certain how best to handle instructing twister which domain you wanted a filter to apply to

File: scripts/twister
Comment (danieldegrasse) [reply scripts/twister]: This has been updated to reflect the actual support present in this PR

**Comment by tgorochowik** [comment]
Removing RISC-V tag as this seems to be accidental here, if you need RISC-V maintainers to review the KConfig change I think it would be better to move this small change to a separate PR.

File: scripts/pylib/twister/twisterlib/runner.py
Comment (hakehuang) [inline scripts/pylib/twister/twisterlib/runner.py:252]: why remove this?

**Review by hakehuang — COMMENTED** [review]
@danieldegrasse  in mcuboot we can also encrypt the boot image, how to add this step in?

File: scripts/pylib/twister/twisterlib/runner.py
Comment (tejlmand) [reply scripts/pylib/twister/twisterlib/runner.py:332]: > We do I believe, because twister does not parse the source application, it parses the generated .config and edt.pickle files in the build folder for an application. This leaves us with the following options:

Thanks for the explanation.
The path you have chosen makes most sense at this stage.

**Comment by danieldegrasse** [comment]
@hakehuang For now, I'd like to just target testing the swap upgrade. I suppose we could encrypt the image during the build process, but my first goal is to verify that the mcuboot swap process works, as this is where platforms like the LPC55s69 will fail

**Comment by hakehuang** [comment]
> @hakehuang For now, I'd like to just target testing the swap upgrade. I suppose we could encrypt the image during the build process, but my first goal is to verify that the mcuboot swap process works, as this is where platforms like the LPC55s69 will fail

@PerMac has a simliar PR before but with different approach, which use yml file to configure the build steps and modify the twister, not in CMake parser. in His PR the encrypt build is supported. but with your sysbuild seems less changes in test application.

**Comment by PerMac** [comment]
@nvlsianpu Could you please check the mcuboot configuration related part?

**Comment by PerMac** [comment]
@danieldegrasse What do we need to expand the scope of platforms able to be used for this tests? E.g. I found that when tried to run on nrf52840dk_nrf52840 I'm getting: 
```
DEBUG   - Flash command: ['ninja', '-C', '/home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot', 'flash']
DEBUG   - ninja: Entering directory `/home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot'

DEBUG   - halted
DEBUG   - run status: nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot error
INFO    - 1/1 nrf52840dk_nrf52840       tests/boot/test_mcuboot/boot.mcuboot               FAILED Device issue (Flash error?) (device 0.050s)
INFO    - /home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot/device.log
ERROR   - ninja: error: unknown target 'flash'
```

**Comment by danieldegrasse** [comment]
@PerMac when I used on device testing with twister to verify the mcuboot sample, I instructed twister to use `west flash`. My final twister call looked like this:
`./scripts/twister -p frdm_k64f -s tests/boot/test_mcuboot/boot.mcuboot --device-testing --device-serial=/dev/ttyS13 --west-flash`

**Review by hakehuang — DISMISSED** [review]
I am OK with this solution.

File: scripts/twister
Comment (tejlmand) [inline scripts/twister:50]: `external` could be a little confusing, as `external` to what.

What about:
```suggestion
    will only be enabled for the main project, and is not supported for
    other projects included by sysbuild.
```

**Review by tejlmand — DISMISSED** [review]
lgtm.

Thanks for taking this first step in integrating sysbuild into twister :+1: 

A minor nit, but nothing blocking.

File: tests/boot/test_mcuboot/testcase.yaml
Comment (carlescufi) [inline tests/boot/test_mcuboot/testcase.yaml:15]: Is there any reason for this list being only 2 boards? There are plenty of other boards that support MCUboot.

Comment (PerMac) [reply tests/boot/test_mcuboot/testcase.yaml:15]: Exactly, `platform_allow` should only been used when there is no way a test can run on anything outside from this list. It mustn't be used to indicate "I tested it on these 2 platforms and it work for me, don't know about the rest".  I expanded it with nrf52840dk_nrf52840 and the test passed. Keeping platform_allow here will prevent others from being able to run this test. Please, propose some other way of filtering if required.

File: tests/boot/test_mcuboot/testcase.yaml
Comment (PerMac) [inline tests/boot/test_mcuboot/testcase.yaml:5]: It an error, should be `type: multi_line`. In the current state the test only looks for the first line and passes just after the first flash.

File: tests/boot/test_mcuboot/testcase.yaml
Comment (PerMac) [inline tests/boot/test_mcuboot/testcase.yaml:14]: what about starting to use sysbuild tag as well? MoOre in the summary comment

**Review by PerMac — CHANGES_REQUESTED** [review]
In general, looks quite promising. Few issues summarized here:
- `one_line` -> `multi_line` in regex (an error)
- usage of `platform_allow` (also an error)
- required `--west-flash` to run which is not used by default (not sure if this is an issue, maybe a proper documentation of this will be enough)
- ~~I opt for adding extra `sysbuild` tag in here~~ _Edit: I realized it is redundant since the test have sysbuild: true. The following stays correct, we need to treat sysbuild tests in an special way by default: using west-flash and make sure no erase is added._ The reasoning: by default in our on-target nightly tests we run twister with `--west-flash=--erase` since there are some tests requiring a clear state of the board in advance. However, it doesn't work with this test, since if we erase the board before every flashing it will screw mcuboot. If a sysbuild tag is used we would be able to exclude such tests from the default call and run them in a one without default erasing. Or twister can recognize this ~~tag~~ option and if used always use west by default to flash and block --erase if used from CLI

I also think this requires a review from someone from mcuboot team (@nvlsianpu @d3zd3z) to check if the required regexes are enough to determine the success and check the test output if it makes sense. I'm adding an output from the test on nrf52840 (after expanding platform_allow and fixing one_line in regex:

```
INFO    - JOBS: 4
INFO    - 1 test scenarios (1 configurations) selected, 0 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
DEBUG   - adding nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot
INFO    - Added initial list of jobs to queue
DEBUG   - Launch process 0
DEBUG   - Launch process 1
DEBUG   - Launch process 2
DEBUG   - Launch process 3
DEBUG   - Running cmake on /home/maciej/zephyrproject/zephyr/tests/boot/test_mcuboot for nrf52840dk_nrf52840
DEBUG   - Building /home/maciej/zephyrproject/zephyr/tests/boot/test_mcuboot using sysbuild
DEBUG   - Calling cmake: /home/maciej/.pyenv/shims/cmake -B/home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot -DTC_RUNID=5874d07816c37c34737395cab01bd709 -DEXTRA_CFLAGS=-Werror '-DEXTRA_AFLAGS=-Werror -Wa,--fatal-warnings' -DEXTRA_LDFLAGS=-Wl,--fatal-warnings -DEXTRA_GEN_DEFINES_ARGS=--edtlib-Werror -GNinja -Sshare/sysbuild -DAPP_DIR=/home/maciej/zephyrproject/zephyr/tests/boot/test_mcuboot -DOVERLAY_CONFIG=/home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot/twister/testsuite_extra.conf -DBOARD=nrf52840dk_nrf52840
DEBUG   - Loaded sysbuild domain data from /home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot/domains.yaml
DEBUG   - Finished building /home/maciej/zephyrproject/zephyr/tests/boot/test_mcuboot for nrf52840dk_nrf52840
DEBUG   - build test: nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot
DEBUG   - Building /home/maciej/zephyrproject/zephyr/tests/boot/test_mcuboot for nrf52840dk_nrf52840
DEBUG   - run test: nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot
DEBUG   - Using serial device /dev/ttyACM0 @ 115200 baud
DEBUG   - Flash command: ['west', 'flash', '--skip-rebuild', '-d', '/home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot']
DEBUG   - DEVICE: *** Booting Zephyr OS build zephyr-v3.1.0-4465-ge013da8716f0  ***
DEBUG   - DEVICE: I: Starting bootloader
DEBUG   - DEVICE: I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
DEBUG   - DEVICE: I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
DEBUG   - DEVICE: I: Boot source: none
DEBUG   - DEVICE: W: Failed reading image headers; Image=0
DEBUG   - DEVICE: E: Unable to find bootable image
DEBUG   - DEVICE: *** Booting Zephyr OS build zephyr-v3.1.0-4465-ge013da8716f0  ***
DEBUG   - DEVICE: I: Starting bootloader
DEBUG   - DEVICE: I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
DEBUG   - DEVICE: I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
DEBUG   - DEVICE: I: Boot source: none
DEBUG   - DEVICE: W: Failed reading image headers; Image=0
DEBUG   - DEVICE: E: Unable to find bootable image
DEBUG   - -- west flash: using runner nrfjprog
-- runners.nrfjprog: mass erase requested
Using board 683205578
-- runners.nrfjprog: Flashing file: /home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot/mcuboot/zephyr/zephyr.hex
Parsing image file.
Verifying programming.
Verified OK.
Enabling pin reset.
Applying pin reset.
-- runners.nrfjprog: Board with serial number 683205578 flashed successfully.
-- west flash: using runner nrfjprog
Using board 683205578
-- runners.nrfjprog: Flashing file: /home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot/swapped_app/zephyr/zephyr.signed.hex
Parsing image file.
Verifying programming.
Verified OK.
Enabling pin reset.
Applying pin reset.
-- runners.nrfjprog: Board with serial number 683205578 flashed successfully.
-- west flash: using runner nrfjprog
Using board 683205578
-- runners.nrfjprog: Flashing file: /home/maciej/zephyrproject/zephyr/twister-out/nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot/test_mcuboot/zephyr/zephyr.signed.hex
Parsing image file.
Verifying programming.
Verified OK.
Enabling pin reset.
Applying pin reset.
-- runners.nrfjprog: Board with serial number 683205578 flashed successfully.

DEBUG   - DEVICE: *** Booting Zephyr OS build zephyr-v3.1.0-4465-ge013da8716f0  ***
DEBUG   - DEVICE: I: Starting bootloader
DEBUG   - DEVICE: I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
DEBUG   - DEVICE: I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
DEBUG   - DEVICE: I: Boot source: none
DEBUG   - DEVICE: I: Swap type: none
DEBUG   - DEVICE: I: Bootloader chainload address offset: 0xc000
DEBUG   - DEVICE: I: Jumping to the first image slot
DEBUG   - DEVICE: *** Booting Zephyr OS build zephyr-v3.1.0-4465-ge013da8716f0  ***
DEBUG   - DEVICE: Launching primary slot application on nrf52840dk_nrf52840
*** Booting Zephyr OS build zephyr-v3.1.0-4465-ge013da8716f0  ***
DEBUG   - DEVICE: I: Starting bootloader
DEBUG   - DEVICE: I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
DEBUG   - DEVICE: I: Secondary image: magic=good, swap_type=0x3, copy_done=0x3, image_ok=0x1
DEBUG   - DEVICE: I: Boot source: none
DEBUG   - DEVICE: I: Swap type: perm
DEBUG   - DEVICE: I: Starting swap using move algorithm.
DEBUG   - DEVICE: *** Booting Zephyr OS build zephyr-v3.1.0-4465-ge013da8716f0  ***
DEBUG   - DEVICE: I: Starting bootloader
DEBUG   - DEVICE: I: Primary image: magic=good, swap_type=0x3, copy_done=0x3, image_ok=0x1
DEBUG   - DEVICE: I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
DEBUG   - DEVICE: I: Boot source: primary slot
DEBUG   - DEVICE: I: Starting swap using move algorithm.
DEBUG   - DEVICE: I: Bootloader chainload address offset: 0xc000
DEBUG   - DEVICE: I: Jumping to the first image slot
DEBUG   - DEVICE: *** Booting Zephyr OS build zephyr-v3.1.0-4465-ge013da8716f0  ***
DEBUG   - DEVICE: Swapped application booted on nrf52840dk_nrf52840
DEBUG   - run status: nrf52840dk_nrf52840/tests/boot/test_mcuboot/boot.mcuboot passed
INFO    - 1/1 nrf52840dk_nrf52840       tests/boot/test_mcuboot/boot.mcuboot               PASSED (device 14.643s)
```

File: tests/boot/test_mcuboot/testcase.yaml
Comment (danieldegrasse) [reply tests/boot/test_mcuboot/testcase.yaml:15]: I agree that `platform_allow` is too restrictive here- however, twister is unable to filter the mcuboot test case in it's current state (since mcuboot requires a boot partition, and devices without that partition will fail to complete CMake configuration)

@tejlmand made me aware of #41302, which would allow twister to filter the MCUBoot sample correctly, by only running some configuration steps. However, as far as I can see twister would need an update to support filtering with only specific cmake modules. I didn't want to over complicate the change set in this PR, so I have not added support for filtering with modules here. My plan was to follow up this PR with one enabling filtering using cmake modules, which will allow the mcuboot test to run on a variety of platforms.

File: tests/boot/test_mcuboot/testcase.yaml
Comment (danieldegrasse) [reply tests/boot/test_mcuboot/testcase.yaml:5]: Thanks for the heads up, I will resolve this on next push.

**Comment by danieldegrasse** [comment]
> usage of platform_allow (also an error)

@PerMac I made a comment above detailing why I am currently using `platform_allow` for the mcuboot sample, but do you want to add support for twister filtering options into this PR? I am OK with adding both changes here if you think it's best, I just wanted to avoid bloating this PR and making the changes to twister difficult to review

File: tests/boot/test_mcuboot/testcase.yaml
Comment (PerMac) [reply tests/boot/test_mcuboot/testcase.yaml:15]: I believe you're more then welcome to add support of #41302 into twister. I was also aware of it but didn't have time yet to add support for it. Since you already dived into twister code and build stages I believe you are well equipped for this task ;) 
IMO (and I know @tejlmand also shares this opinion) it would be super useful in twister: allowing us to save time building test which will end skipped in the end and making this filtration available without full cmake (which can fail).

**Comment by PerMac** [comment]
@danieldegrasse I believe it should definitely be a separate PR. I can live with temporarily having this "platform_allow" and not block the PR because of it :)

**Comment by danieldegrasse** [comment]
@PerMac Sounds good-I'll make the remaining requested changes to this PR, and follow this PR up with another to enable filtering.

**Comment by d3zd3z** [comment]
> For now, I'd like to just target testing the swap upgrade. I suppose we could encrypt the image during the build process, but my first goal is to verify that the mcuboot swap process works, as this is where platforms like the LPC55s69 will fail

Is the intent to demonstrate that the LPC55S69 fails with swap? There is development work that should eventually implement swap for LPC55S69 (but it won't be swap scratch or swap move, but a new one), but this is fairly slow going.

Also, when you say "swap" are you meaning "swap scratch" or "swap move". I'm trying to get to where nobody uses swap scratch any more, because swap move is better, and should be supported on any target we support in Zephyr.

**Comment by danieldegrasse** [comment]
@d3zd3z What I want to verify with the MCUBoot test is that the default configuration for a given board when running with MCUBoot supports the following:

- chainloading a Zephyr executable
- starting the swap process to a new image
- swapping to that new image, and booting it as the confirmed image

I agree that targets should support swap move over swap scratch, but I think we should implement a baseline "default configuration" test first, then tests can be added that explicitly require the swapping processes to use swap move

**Comment by danieldegrasse** [comment]
@PerMac this PR has been updated with documentation to mention that west or the `hardware_map` option must be used to flash applications when using sysbuild, and the `multi_line` regex error in the test has been fixed.

File: scripts/pylib/twister/twisterlib/runner.py
Comment (mbolivar-nordic) [inline scripts/pylib/twister/twisterlib/runner.py:394]: whitespace nit

File: tests/boot/test_mcuboot/sysbuild.cmake
Comment (mbolivar-nordic) [inline tests/boot/test_mcuboot/sysbuild.cmake:6]: Why is the `if` necessary, if the point of the sample is to test this?

File: tests/boot/test_mcuboot/sysbuild.cmake
Comment (mbolivar-nordic) [inline tests/boot/test_mcuboot/sysbuild.cmake:7]: Why not just stick this in the swapped_app/prj.conf file?

**Comment by PerMac** [comment]
@danieldegrasse I think my comment didn't reflect what I was thinking, sorry fot that. Let me be more precise:
- Since sysbuild can only be tested by using west, maybe `options.west_flash` should be set by default for such tests? I.e. the option to use west for flashing (`options.west_flash`) should be turned on in twister for sysbuild tests? What's more, I think twister should remove the `--erase` from west parameters if a user added them in the CLI (e.g. using `--west-flash=--erase`), because using `--erase` screws mcuboot. @nashif @mbolivar-nordic What do you think about such approach? Reasoning: we use `--west-flash=--erase` by default in our nightly on-target plan since some tests (e.g. tfm) can only pass with it but there is no logic to force erase from the test level. If twister won't remove `--erase` in such case, we would have to split the call so sybuild tests are run independently with different command. 
- If it is decided that the workflow I am proposing above shouldn't be used, then document the need for west for flashing (you did it already) and add a filter skipping sysbuild tests if west was not chosen + add a clear reason why the test was skipped.

File: doc/develop/test/twister.rst
Comment (mniestroj) [inline doc/develop/test/twister.rst:370]: This should not be under `harness_config`.

File: scripts/pylib/twister/twisterlib/runner.py
Comment (mniestroj) [inline scripts/pylib/twister/twisterlib/runner.py:264]:
```suggestion
                f'-S{canonical_zephyr_base}/share/sysbuild',
```
to make it work from outside of Zephyr root directory.

**Review by mniestroj — CHANGES_REQUESTED** [review]

**Comment by mbolivar-nordic** [comment]
> @nashif @mbolivar-nordic What do you think about such approach?

Fine by me

**Review by nvlsianpu — COMMENTED** [review]
I reserved some time to review this PR next week (19.09-23.09)

File: tests/boot/test_mcuboot/swapped_app/Kconfig
Comment (nordicjm) [inline tests/boot/test_mcuboot/swapped_app/Kconfig:7]: This does not work for devices where the secondary slot it not part of the primary flash, e.g. in an external flash.

File: tests/boot/test_mcuboot/testcase.yaml
Comment (nvlsianpu) [reply tests/boot/test_mcuboot/testcase.yaml:15]: I'm not convinced that some more sophisticated mechanism will always works. MCUboot basically requires proper flash partitioning, flash driver which supports write-block-size <= 32 B, supported by it architecture (for instance CortexM3...). Even that platform fulfil all these requirement MCUboot might be not the case for the board (for instance some flash memories are really weird - therefore incompatible, different bootloader is expected etc.)

File: tests/boot/test_mcuboot/swapped_app/Kconfig
Comment (nvlsianpu) [reply tests/boot/test_mcuboot/swapped_app/Kconfig:7]: So such boards shouldn't be used as build targets.

Comment (nordicjm) [reply tests/boot/test_mcuboot/swapped_app/Kconfig:7]: That seems like a big limitation. Would it not be better to have a Kconfig option to generate for the secondary slot instead? The list of devices might not be large at present but it is going to only grow with time (devices where slot1 and slot0 differ):
```
arm/usb_kw24d512/usb_kw24d512.dts
arm/rm1xx_dvk/rm1xx_dvk.dts
arm/pinnacle_100_dvk/pinnacle_100_dvk.dts
arm/lpcxpresso55s36/lpcxpresso55s36.dts
arm/lpcxpresso55s28/lpcxpresso55s28.dts
arm/lpcxpresso55s16/lpcxpresso55s16_common.dtsi
arm/lpcxpresso55s06/lpcxpresso55s06_common.dtsi
arm/lpcxpresso54114/lpcxpresso54114_m4.dts
arm/frdm_kw41z/frdm_kw41z.dts
arm/disco_l475_iot1/disco_l475_iot1.dts
arm/bt610/bt610.dts
arm/bl5340_dvk/bl5340_dvk_cpuapp_common.dts
```

Comment (d3zd3z) [reply tests/boot/test_mcuboot/swapped_app/Kconfig:7]: It seems that the second slot being on an external device isn't tested much because dev boards usually don't have external flash. But, it does seem common in production. I think it is an important case to consider.

Comment (nordicjm) [reply tests/boot/test_mcuboot/swapped_app/Kconfig:7]: > It seems that the second slot being on an external device isn't tested much because dev boards usually don't have external flash. But, it does seem common in production. I think it is an important case to consider.

I had a quick grep through the boards and it seems 63 boards have QSPI entries in their boards files, 63/500 is 13% of all supported boards, which isn't a small number.

Comment (d3zd3z) [reply tests/boot/test_mcuboot/swapped_app/Kconfig:7]: To clarify, "isn't tested very much" by things like our testing infrastructure. I'm sure there are lots of targets that do this, and that it is pretty common in production.

**Comment by nashif** [comment]
>  What's more, I think twister should remove the --erase from west parameters if a user added them in the CLI (e.g. using --west-flash=--erase), because using --erase screws mcuboot. @nashif @mbolivar-nordic What do you think about such approach?

I do not think it is a good idea for twister to remove anything that was added by the caller on the command line. If this option is not supported, twister needs to error out. Removing options silently will just cause problems and caller will be left wondering what happened...

**Comment by nordicjm** [comment]
The erase/recover part needs to be fixed or blocked entirely, it's not acceptable that doing a twister erase or `west flash --recovery` programs 3 different files and recovers the module prior to each leaving it useless https://github.com/zephyrproject-rtos/zephyr/issues/50421 there shouldn't need to be different flashing commands for if you are using sysbuild vs not using sysbuild to do the same thing, adding sysbuild should be transparent to the user in this regard.

**Comment by danieldegrasse** [comment]
@nashif I agree that silently ignoring command line parameters isn't the best way to handle this issue, but the alternative seems to be breaking @PerMac's nightly testing flow. @PerMac, you mentioned that there are only some subset of tests (I know you called out TFM specifically) that require an erase. Maybe the best option here is to add the ability to encode a requirement to erase the flash before flashing a sample into twister?

File: tests/boot/test_mcuboot/swapped_app/Kconfig
Comment (danieldegrasse) [reply tests/boot/test_mcuboot/swapped_app/Kconfig:7]: > This does not work for devices where the secondary slot is not part of the primary flash, e.g. in an external flash.

What would the required steps here be in order to generate an image for the secondary flash? The image needs to be build targeting the primary slot, then the programming tool (pyocd, jlink,...) needs to be instructed to load the image at a given offset, correct? Or if the image cannot be loaded using jlink or another west runner, what would be the method to flash it?

Perhaps the work in #50393 could be leveraged to pass a load address to the runner?

Comment (nordicjm) [reply tests/boot/test_mcuboot/swapped_app/Kconfig:7]: For nordic parts, if the secondary slot it in QSPI then it is memory mapped, 0x12000000 for nRF52840 and 0x10000000 for nRF5340, this can be programmed directly to the QSPI flash using nrfjprog, I would assume other flash runners would not support programming the QSPI. There is offset given to nrfjprog with e.g. a bin file, it needs a properly addressed hex file. For SPI flash on other nordic parts, e.g. nRF52832, they cannot be flashed from nrfjprog, flashing of such a flash would be down to whomever creates the board (it would need to be a custom flash runner)

**Comment by PerMac** [comment]
@nashif @danieldegrasse I guess adding a feature where a test could order an erase as a part of the setup would be nice. I agree the most transparent would be producing an error when sysbuild test is called with not compatible CLI arguments. I don't want to sound like "it dosn't work with my way of running twister, so I will block it", just wanted to show, that sysbuild tests require extra "guarding logic", since they require a specific workflow and some indirect and harder to follow errors can pop out. As the easiest solution for now I see adding a "sysbuild" tag to sysbuild tests/samples. I (and ofc others) could then split the workflow calling twister twice, with  `--exclude-tag`\`--tag` args

**Comment by danieldegrasse** [comment]
> I (and ofc others) could then split the workflow calling twister twice, with --exclude-tag`--tag` args

@PerMac Ok, if you can still use this implementation within your tests by running sysbuild tests separately, I will update this PR so that twister will skip the test if called without `options.west_flash` set to true, or with `--erase` passed to the west options. This way existing test runs will not break, and users can enable sysbuild tests by making sure their twister runs satisfy these flags

File: scripts/pylib/twister/twisterlib/runner.py
Comment (danieldegrasse) [reply scripts/pylib/twister/twisterlib/runner.py]: Fixed in latest push

File: tests/boot/test_mcuboot/sysbuild.cmake
Comment (danieldegrasse) [reply tests/boot/test_mcuboot/sysbuild.cmake]: It's not- I've removed this

File: tests/boot/test_mcuboot/sysbuild.cmake
Comment (danieldegrasse) [reply tests/boot/test_mcuboot/sysbuild.cmake]: For `CONFIG_BOOTLOADER_MCUBOOT`, the `prj.conf` file can be used, and I've moved that KConfig there. `CONFIG_MCUBOOT_SIGNATURE_KEY_FILE` does need to be set like this though, to ensure it takes the same value as the main application and MCUBoot get.

File: doc/develop/test/twister.rst
Comment (danieldegrasse) [reply doc/develop/test/twister.rst]: Fixed, thanks for catching this

File: scripts/pylib/twister/twisterlib/runner.py
Comment (danieldegrasse) [reply scripts/pylib/twister/twisterlib/runner.py]: fixed in latest push

**Comment by danieldegrasse** [comment]
@PerMac  @mniestroj I've updated this PR based on your feedback. The current state is as follows:

- Sysbuild support parses the KConfig and DTS of the "main" application being built with sysbuild for filtering purposes
- `platform_allow` is still in use- see the discussion above for the future solution using PR #41302
- When twister is run with `--west-flash=--erase`, or without `--west-flash` (both unsupported settings) then tests requiring sysbuild support will be skipped, and a warning will be printed (@nordicjm from one of your prior comments, it appears that `--west-flash=--recover` may also be an unsupported flag?)
- Documentation has been updated to indicate the above behavior when a test requires sysbuild

**Comment by PerMac** [comment]
Thanks for the changes, works as described

**Review by PerMac — APPROVED** [review]

**Review by hakehuang — APPROVED** [review]

**Review by mniestroj — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**
